File: CPP/Problem_118.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <functional>
using namespace std;

using u64 = unsigned long long;

static u64 mod_mul(u64 a, u64 b, u64 mod) {
    u64 res = 0;
    a %= mod;

    while (b) {
        if (b & 1)
            res = (res + a) % mod;

        a = (a << 1) % mod;
        b >>= 1;
    }

    return res;
}
static u64 mod_pow(u64 a, u64 d, u64 mod) {
    u64 r = 1;
    while (d) {
        if (d & 1) r = mod_mul(r, a, mod);
        a = mod_mul(a, a, mod);
        d >>= 1;
    }
    return r;
}

// Deterministic Miller-Rabin for 32-bit-ish range (we're < 1e9)
static bool isPrime(u64 n) {
    if (n < 2) return false;
    for (u64 p : {2ULL,3ULL,5ULL,7ULL,11ULL,13ULL,17ULL,19ULL,23ULL,29ULL,31ULL,37ULL}) {
        if (n % p == 0) return n == p;
    }
    u64 d = n - 1, s = 0;
    while ((d & 1) == 0) { d >>= 1; s++; }

    auto witness = [&](u64 a)->bool{
        if (a % n == 0) return true;
        u64 x = mod_pow(a, d, n);
        if (x == 1 || x == n - 1) return true;
        for (u64 i = 1; i < s; i++) {
            x = mod_mul(x, x, n);
            if (x == n - 1) return true;
        }
        return false;
    };

    for (u64 a : {2ULL,3ULL,5ULL,7ULL,11ULL}) {
        if (!witness(a)) return false;
    }
    return true;
}

static const int ALL = (1<<9) - 1;
static vector<vector<int>> primesByMask(ALL+1);

static vector<int> digitsFromMask(int mask) {
    vector<int> d;
    for (int i = 0; i < 9; i++) if (mask & (1<<i)) d.push_back(i+1);
    return d;
}

static vector<int> buildPrimesForMask(int mask) {
    vector<int> digs = digitsFromMask(mask);
    int sum = 0;
    for (int x: digs) sum += x;
    if (sum % 3 == 0 && !(digs.size() == 1 && digs[0] == 3)) return {};

    sort(digs.begin(), digs.end());
    vector<int> used(digs.size(), 0);
    unordered_set<int> res;
    res.reserve(128);

    function<void(int,int)> backtrack = [&](int cur, int depth){
        if (depth == (int)digs.size()) {
            if (digs.size() > 1) {
                int last = cur % 10;
                if (last % 2 == 0 || last == 5) return;
            }
            if (isPrime((u64)cur)) res.insert(cur);
            return;
        }
        for (int i = 0; i < (int)digs.size(); i++) {
            if (used[i]) continue;
            used[i] = 1;
            backtrack(cur*10 + digs[i], depth+1);
            used[i] = 0;
        }
    };

    backtrack(0, 0);
    vector<int> out(res.begin(), res.end());
    sort(out.begin(), out.end());
    return out;
}

// memo: remaining_mask + last_val hashed
struct Key {
    int mask;
    int last;
    bool operator==(Key const& o) const { return mask==o.mask && last==o.last; }
};
struct KeyHash {
    size_t operator()(Key const& k) const {
        return (size_t)k.mask * 1000003u ^ (size_t)k.last;
    }
};

static unordered_map<Key, long long, KeyHash> memo;

static long long dfs(int remaining, int lastVal) {
    if (remaining == 0) return 1;
    Key key{remaining, lastVal};
    auto it = memo.find(key);
    if (it != memo.end()) return it->second;

    long long total = 0;
    for (int sub = remaining; sub; sub = (sub - 1) & remaining) {
        auto &plist = primesByMask[sub];
        if (plist.empty()) continue;
        auto pos = upper_bound(plist.begin(), plist.end(), lastVal);
        for (auto jt = pos; jt != plist.end(); ++jt) {
            total += dfs(remaining ^ sub, *jt);
        }
    }
    memo.emplace(key, total);
    return total;
}

int main() {
    for (int m = 1; m <= ALL; m++) primesByMask[m] = buildPrimesForMask(m);
    cout << dfs(ALL, 0) << "\n"; // 44680
    return 0;
}

File: Python/Problem_118.py
from functools import lru_cache

ALL = (1 << 9) - 1  # digits 1..9

def is_prime(n: int) -> bool:
    if n < 2:
        return False
    small = (2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37)
    for p in small:
        if n % p == 0:
            return n == p

    # Miller-Rabin for 32-bit range is overkill but fast in Python
    d = n - 1
    s = 0
    while d % 2 == 0:
        s += 1
        d //= 2

    def check(a: int) -> bool:
        x = pow(a, d, n)
        if x == 1 or x == n - 1:
            return True
        for _ in range(s - 1):
            x = (x * x) % n
            if x == n - 1:
                return True
        return False

    for a in (2, 3, 5, 7, 11):
        if a % n == 0:
            continue
        if not check(a):
            return False
    return True

def digits_from_mask(mask: int):
    # bit0 -> digit 1, bit8 -> digit 9
    return [i + 1 for i in range(9) if (mask >> i) & 1]

def primes_for_mask(mask: int):
    digs = digits_from_mask(mask)
    # digit-sum mod 3 pruning (only single-digit 3 survives)
    s = sum(digs)
    if s % 3 == 0 and not (len(digs) == 1 and digs[0] == 3):
        return []

    res = set()
    used = [False] * len(digs)
    digs.sort()

    def backtrack(cur: int, depth: int):
        if depth == len(digs):
            # quick last digit filter
            if len(digs) > 1:
                last = cur % 10
                if last % 2 == 0 or last == 5:
                    return
            if is_prime(cur):
                res.add(cur)
            return

        for i in range(len(digs)):
            if used[i]:
                continue
            # avoid duplicate permutations when digits repeat (they don't here, but keep it clean)
            used[i] = True
            backtrack(cur * 10 + digs[i], depth + 1)
            used[i] = False

    backtrack(0, 0)
    return sorted(res)

# precompute prime lists per mask
PRIMES = [[] for _ in range(ALL + 1)]
for m in range(1, ALL + 1):
    PRIMES[m] = primes_for_mask(m)

@lru_cache(maxsize=None)
def dfs(remaining_mask: int, last_val: int) -> int:
    if remaining_mask == 0:
        return 1

    total = 0
    sub = remaining_mask
    while sub:
        plist = PRIMES[sub]
        if plist:
            # iterate only values > last_val (plist sorted)
            # manual bisect to avoid import, because why not
            lo, hi = 0, len(plist)
            while lo < hi:
                mid = (lo + hi) // 2
                if plist[mid] <= last_val:
                    lo = mid + 1
                else:
                    hi = mid
            for v in plist[lo:]:
                total += dfs(remaining_mask ^ sub, v)
        sub = (sub - 1) & remaining_mask

    return total

def solve():
    return dfs(ALL, 0)

if __name__ == "__main__":
    print(solve())  # 44680
